File: DSA_using_C/Data Structures/Stack/queue_using_array.c
#include <stdio.h>

#define SIZE 100

struct Queue {
    int myList[SIZE];
    int size;
    int front;
    int rear;
};

void initializeQueue(struct Queue *queue, int size) {
    int i;
    for (i = 0; i < size; i++) {
        queue->myList[i] = i;
    }
    queue->size = size;
    queue->front = -1;
    queue->rear = -1;
}

void enqueue(struct Queue *queue, int num) {
    if (queue->rear == queue->size - 1) {
        printf("Queue is already full\n");
        return;
    } else {
        queue->rear++;
        queue->myList[queue->rear] = num;
    }
}

int dequeue(struct Queue *queue) {
    if (queue->rear == -1) {
        printf("Queue is already empty\n");
        return -1; // Assuming -1 represents an error or an invalid value
    } else {
        int temp = queue->myList[0];
        for (int i = 0; i < queue->rear; i++) {
            queue->myList[i] = queue->myList[i + 1];
        }
        queue->rear--;
        return temp;
    }
}

int isEmpty(struct Queue *queue) {
    return (queue->rear == -1);
}

int isFull(struct Queue *queue) {
    return (queue->rear == queue->size - 1);
}

int currSize(struct Queue *queue) {
    return (queue->rear + 1);
}

void print(struct Queue *queue) {
    printf("Dequeue <- ");
    for (int i = 0; i <= queue->rear; i++) {
        printf("%d | ", queue->myList[i]);
    }
    printf("<- Enqueue\n");
}

int main() {
    struct Queue queue;
    initializeQueue(&queue, 100);

    for (int i = 1; i < SIZE; i++) {
        enqueue(&queue, i);
    }

    printf("Queue is full: %s\n", isFull(&queue) ? "true" : "false");
    printf("Queue size: %d\n", currSize(&queue));
    print(&queue);

    for (int i = 1; i <= 7; i++) {
        printf("The element removed is: %d\n", dequeue(&queue));
    }

    printf("Queue is empty: %s\n", isEmpty(&queue) ? "true" : "false");
    printf("Queue size: %d\n", currSize(&queue));

    return 0;
}

File: DSA_using_python/Data_Structures/queue_using_array.py
class Queue:
    myList=list()
    size=0
    front=-1
    rear=-1

    def __init__(self,size):
        self.myList.extend(list(range(0,size)))
        self.size = size

    def isEmpty(self):
        if self.rear==-1 : return True
        else: return False

    def isFull(self):
        if self.rear==self.size-1: return True
        else: return False

    def enqueue(self, num):
        if self.isFull():
            print("Queue is already full")
            return
        else:
            self.rear+=1
            self.myList[self.rear]=num

    def dequeue(self):
        if self.isEmpty():
            print("Queue is already empty")
            return
        else:
            temp = self.myList[0]
            #first idx will be removed, we need to shift all elements to the left
            for i in range(0,self.rear):
                self.myList[i]=self.myList[i+1]
            self.rear-=1
            return temp

    def currSize(self):
        return self.rear+1

    def print(self):
        print("Dequeue <- ",end='')
        for i in range(0,self.currSize()):
            print(self.myList[i],"|",end=' ')
        print("<- Enqueue")

#Below is some driver code
queue = Queue(25)
for i in range(1,25):
    queue.enqueue(i)
print("Queue is full: {}".format(queue.isFull()))
print("Queue size: {}".format(queue.currSize()))
queue.print()
for i in range(1,8):
    print("The element removed is: {}".format(queue.dequeue()))
print("Queue is empty: {}".format(queue.isEmpty()))
print("Queue size: {}".format(queue.currSize()))
queue.print()
                           
